### ISSUE TITLE
Script Node

### ISSUE THREAD

**Issue by ly29** [Sverchok/Sverchok#26]
![script_baby](https://cloud.githubusercontent.com/assets/6241382/22639845/4f2cfca8-ec50-11e6-9027-43da1373c49d.gif)
So just started working on it.

Barely there but functional in the sense that you can load a script and it works.

Todo: (or what does not work)
- [x] Reset
- [x] Editing
- [x] Reload
- [x] UI
- [x] Reload on open
- [ ] Reload from TextEditor
- [x] First line of text must be empty
- [x] Text file must be valid python identifier (in other words allow ```file.py``` and similar things)

**Comment by ly29** [comment]
So a very basic, somewhat rough version working, no custom drawing or custom properties but as a basic thing for testing and developing node functions it works.

**Comment by ly29** [comment]
Basic syntax as follows, use ```@node_script``` to designate the function you want to test. No extra options allowed right now. The function should *behave* exactly as full node, including load on start up as long as the text file exist.


```python
@node_script
def add(a: Int = 0, b: Int = 1) -> Number
    return a + b
```

Caveats
- The text file must have a name that is valid python module identifier. 
- First line of text file must be empty
- Very rough ui
- No special drawing
- No custom properties
- no multi mode
- etc.

**Comment by ly29** [comment]
This should speed up development speed

**Comment by ly29** [comment]
Simple example

![skarmavbild 2017-02-06 kl 10 53 05](https://cloud.githubusercontent.com/assets/6241382/22642303/7baddc34-ec5a-11e6-8c0a-554e3ddf04c3.png)

Corrected example
```python

import numpy as np

@node_script
def add(verts: Vertices = None, scale : Float = 1.0) -> Vertices:
    s= scale[0]
    return verts * np.array([s, s, s, 0])
```

**Comment by portnov** [comment]
> s = scale[0]

looks weird :)

**Comment by ly29** [comment]
Yeah, there is a reason for that, sadly and it is result of an inconsistency  between data and sockets.

Since we are using numpy I decided to wrap the socket data to ```np.array([value])``` to make it consistent with incoming data from another  node, which has a side effect in this case.

**Comment by ly29** [comment]
Since if we would drop a list of then different scales into the node we should get 10 different list of data.

**Comment by portnov** [comment]
In this case we will get a problem with specific nodes needing vectorization again. In fact, in your example node author have to make a decision of whether to get [0] or iterate by the array or...?
I think this decision must be moved into node settings or core code.

**Comment by zeffii** [comment]
```python

import numpy as np
import bmesh

def rxdata_from_bmesh(bm):
    v = np.array([v.co[:] for v in bm.verts])
    # e = np.array([[i.index for i in e.verts] for e in bm.edges])
    # p = np.array([[i.index for i in p.verts] for p in bm.faces])
    return v #, e #, p    

@node_script
def sn_icosphere(subdiv: Int = 2, diam: Float = 1.0) -> Vertices:
    bm = bmesh.new()
    bmesh.ops.create_icosphere(bm, subdivisions=4, diameter=diam, calc_uvs=False)
    return rxdata_from_bmesh(bm)


```

**Comment by ly29** [comment]
While I agree that my example is deeply flawed there is no clean solution possible for matching every possible case. 

I was somewhat hurried this morning and didn't make the choice how the scale should be understood, if a scaling each set of vertices separately or if a the scale should be understood per vertex. Both are possible understandings, and it is and issue to be resolved but not an impossible problem with the base system, but a choice that has to be made, and clearly shown.

**Comment by zeffii** [comment]
i like to have control over how to deal with nested input, that wasn't a weakness of 0.59 imo..

**Comment by ly29** [comment]
The node should have control, but ideally by declaring what type of data it wants.

**Comment by ly29** [comment]
Anyway that is more of type issue and not a script node issue

**Comment by portnov** [comment]
Imho the most flexible way would be to move this matching logic into node settings (common for all nodes). Maybe node could declare what kind of matching it wants by default. But the user could redefine it.

**Comment by ly29** [comment]
the matching is outside of the node, the question is of what to match. 
that is what the type system tries to solve, there are some problems there admittedly, but they are in fact not that big or impossible to solve.

**Comment by portnov** [comment]
>  the question is of what to match.

My proposal is to move it to node settings. In current sverchok we have "list match" node; imho it will be more usefull and probably better from performance side to have the same settings in each node.

**Comment by zeffii** [comment]
```python


import numpy as np
import bmesh

def rxdata_from_bmesh(bm):
    v = np.array([v.co[:] for v in bm.verts])
    e = np.array([[i.index for i in e.verts] for e in bm.edges])
    # p = np.array([[i.index for i in p.verts] for p in bm.faces])
    return v, e # , p    

@node_script
def sn_icosphere(subdiv: Int = 2, diam: Float = 1.0) -> ([Vertices], [Edges]):
    bm = bmesh.new()
    bmesh.ops.create_icosphere(bm, subdivisions=4, diameter=diam, calc_uvs=False)
    return rxdata_from_bmesh(bm)   # even if this is wrapped in a list
```
gives an error
```python
  File "svrx\nodes\output\mesh_out.py", line 37, in stop
    obj = make_bmesh_geometry(verts[:, :3], edges, faces,
IndexError: too many indices for array
```
being that all the faces are tris, that could be done with flat storage out of  the box..

**Comment by ly29** [comment]
There is converter for Sverchok style data in smesh in util.

**Comment by ly29** [comment]
Hmm also vertices should be n,4 in shape. Will check in detail tomorrow. Not sure what goes wrong, I think that should work

**Comment by zeffii** [comment]
it makes the vertices, but as soon as i try to connect Edges or Faces i get that error

**Comment by zeffii** [comment]
oh I see, then rename it to SvPolygons ?

**Comment by ly29** [comment]
Check the shapes of the arrays also

**Comment by zeffii** [comment]
it makes sense to do a proper `svrxdata_from_bm` soon.

**Comment by ly29** [comment]
Absolutely

**Comment by zeffii** [comment]
i'll start reading the code closer now :)

**Comment by ly29** [comment]
Ask if wonder anything

**Comment by ly29** [comment]
Still many parts are a bit sketchy

**Comment by ly29** [comment]
@zeffii 
So ```rxdata_from_bm``` exists now

```python

import bmesh
from svrx.util.geom import vectorize
from svrx.util.mesh import rxdata_from_bm

@vectorize
def make_icosphere(subdiv, diam):
    bm = bmesh.new()
    bmesh.ops.create_icosphere(bm, subdivisions=subdiv, diameter=diam, calc_uvs=False)
    return rxdata_from_bm(bm)

@node_script
def sn_icosphere(subdiv: Int = 2, diam: Float = 1.0) -> ([Vertices], [Edges], [Faces]):
    return list(make_icosphere(subdiv, diam))    
```

This is how a generator functions looks right now.  As stated the problem is that there is no type that corresponds to a singular number but ```Float```, ```Int``` corresponds to a list of numbers and the smallest part is an ```np.array``` with corresponding shape ```(n,)``` .

Is this to much complexity for the generator functions? It is a very common pattern to produce complexity and something that should be done more clearly. It could easily be wrapped into a decorator function. ```@vectorize``` basically does it already but use ```yield``` causing the ```return list(...)``` **or** we could create a type that actually corresponds to a singular number, which makes a lot of sense.

Many functions do not fall into the above generator pattern but need the whole list, in broadcastable format  for many cases.

**Comment by ly29** [comment]
For the how part we need both node and settings since one might want match data and reuse it in different nodes.

**Comment by zeffii** [comment]
that looks fine, remember we talked about a vectorize decorator early on. https://github.com/nortikin/sverchok/issues/105

**Comment by ly29** [comment]
I should reread that I think :)

**Comment by zeffii** [comment]
this is really quite cool @ly29 - very nice work.

I have a suspicion that the regular sverchok monad could "host" a rx layout?

**Comment by ly29** [comment]
Making a monad is basically composing a layout into a node function, so yes it could be done with some conventions about how to flatten data at the end.

**Comment by zeffii** [comment]
btw: what's the deal with the requirement to have an empty line at the top of a `sn`? this `standard header` injection?

**Comment by ly29** [comment]
Yeah, that is very small todo, basically inserting an extra ```;``` into the end of the standard header.

Also apart from the a problem with how they appear in the menu, that could be fixed, the same mechanism can support "full" nodes using ```@node_func``` and ```@stateful``` easily.

**Comment by zeffii** [comment]
also.. how does one limit for example ` subdiv: Int = 2, `   to a maximum of say...5 on the ui/socket

**Comment by ly29** [comment]
>  how does one limit the this subdiv: Int = 2, to a maximum of say...5

That is a good question which I am thinking about and have some different solutions to. 

For the property one could do various things.
- Create a node property and expose in the socket, I rather not go this way.
- Create custom sockets for each case, not a such a good idea either
- Create controllable socket properties with either class or per node limits

I think the third option is the way to go using ```get/set``` functions on the exposed property

**Comment by ly29** [comment]
Still one would have confirm that numbers fall into the appropriate range inside of the node and/or the system. Here we can go both ways and it ties together with #24

**Comment by zeffii** [comment]
`   subdiv: IntP(max=5) = 2,  `  ?

**Comment by ly29** [comment]
> btw: what's the deal with the requirement to have an empty line at the top of a sn?

Yes that goes together with injecting the standard header, however. Actually an extra ```;``` there should fix that I see now. I wanted to keep the line number for errors.

https://github.com/Sverchok/Sverchok/blob/master/core/script.py#L49-L63


> `   subdiv: IntP(max=5) = 2,  `  ?

Something like that the syntax would become yes. ```UInt(max=5) = 2``` would be nice

**Comment by zeffii** [comment]
> Still one would have confirm that numbers fall into the appropriate range inside of the node and/or the system.

yeah, first line of defence is nice to have --   (this may change)

start thinking about design docs for users:
- When exposing variables to the node and its users here are a few tips
    - Figure out if your script's algorithm needs some min / max limits on certain properties.
    - If you know that values beyond a certain range will take very long to process, use a max on the socket, and test for the max inside your code. The socket's max only has effect on the value when the value comes from the slider.

**Comment by zeffii** [comment]
I know it's early, but : https://github.com/Sverchok/Sverchok/wiki/ScriptNode-RX-cookbook

**Comment by zeffii** [comment]
https://github.com/Sverchok/Sverchok/wiki/ScriptNode-RX-cookbook#torus 
![image](https://cloud.githubusercontent.com/assets/619340/22688905/882bf91c-ed2e-11e6-8ae9-fd4a9013a672.png)

**Comment by ly29** [comment]
Nice work!

**Comment by ly29** [comment]
Nice work!

**Comment by zeffii** [comment]
Sn takes label from nodescript decorated function and not text-datablok name, i have mixed feelings about this

**Comment by ly29** [comment]
Ideally the text name and function name should align, but to be clear perhaps the file name could be drawn in ```draw_buttons```

**Comment by ly29** [comment]
Added from script to node instructions in the wiki. That could easily be turned into a script...

**Comment by zeffii** [comment]
let's be cheeky and disallow 'Text' as a filename :)

**Comment by zeffii** [comment]
I think i'll temporarily add svrx trees to the sverchok operator in TextEditor. but eventually i'm sure we can add a kb combo that only triggers if a certain node tree is active/last used.

**Comment by zeffii** [comment]
how about `draw_buttons_ext` gets a way to call some custom_draw function too, i'd like something like

```python
def draw_buttons_ext(self, context, layout):
    layout_draw_weblink('SvRx Cookbook')   #
    layout_draw_simple_templates()    # maybe 10 scipts and a signature file with multiple templates.
    if self.custom_ext:
         self.custom_draw_buttons_ext(context, layout)
```

**Comment by ly29** [comment]
I am for easily accessible and well commented templates/examples

**Comment by zeffii** [comment]
right.. maybe add the link to the cookbook inside the _signature templates_ file , no need to draw weblink in `ext`.

**Comment by ly29** [comment]
Cook something together :)

**Comment by zeffii** [comment]
ok.. for now the templates will have to drop the .py extention  :)

**Comment by ly29** [comment]
Dot py and basiclly any name should now via name mangling.
Unless you have file called say ```script.py``` and another called ```scriptpy``` but, honestly, just don't do that.

**Comment by ly29** [comment]
Will add load error message using error module.

**Comment by ly29** [comment]
![syntaxerror](https://cloud.githubusercontent.com/assets/6241382/23204051/4fdc4d88-f8e5-11e6-8083-65218d8c6c51.PNG)

**Comment by ly29** [comment]
Would be nice if one could highlight the error as well in the text editor

**Comment by zeffii** [comment]
we can put the cursor on the line 
```
# t is a reference to the text object.
t.current_line_index = line_index
```

and `bpy.ops.text.select_line()` is as far as i'm aware the only way to select text from python.

**Comment by ly29** [comment]
This I guess but I rather not go there...
https://docs.blender.org/api/blender_python_api_2_76_2/bpy.types.SpaceTextEditor.html?highlight=region_location_from_cursor#bpy.types.SpaceTextEditor.region_location_from_cursor

**Comment by zeffii** [comment]
where's the emoticon for vomit.

**Comment by zeffii** [comment]
i motion to resist doing that, and lobby instead for more TextEditor accessors in 2.8.

**Comment by ly29** [comment]
>where's the emoticon for vomit

🤢 

> i motion to resist doing that

👍

**Comment by zeffii** [comment]
i'm missing something obvious here... but how do i make a controller script..
```
from svrx.typing import Int, ColorP
import bpy

@node_script
def set_color(edge_col: ColorP(default=(0.3, 0.2, 0.5)) = Required, num: Int = 2) -> Int:
    
    nodes = bpy.data.node_groups['NodeTree.002'].nodes
    
    nodes['View bm GL'].edge_color = edge_col
    nodes['View bm GL.001'].edge_color = edge_col
    nodes['View bm GL.002'].edge_color = edge_col
    
    return 2
```
here the output is not wanted.. and also the body doesn't get executed, also not sure why it's not accepting ColorP .   a little bit lost

**Comment by ly29** [comment]
Scriptnode doesn't accept properties yet, didn't want get into the whole creating classes thing there yet.

This works
```python

import bpy

@node_script
def set_color(edge_col: Color() = Required):
    edge_col = edge_col[:3]
    nodes = bpy.data.node_groups['NodeTree.002'].nodes
    
    nodes['View bm GL'].edge_color = edge_col
    nodes['View bm GL.001'].edge_color = edge_col
    nodes['View bm GL.002'].edge_color = edge_col

```

![script_control](https://cloud.githubusercontent.com/assets/6241382/23343702/2acef390-fc70-11e6-92b8-6d948dbaebca.PNG)

**Comment by ly29** [comment]
I doesn't stop them from existing but doesn't create them either

**Comment by ly29** [comment]
Also nodes doesn't get evaluated unless they are connected.

one can have full node_func stateful in the text editor and they should appear under User in the menu. Just reference the file with a script node and it would work.
